test/test_gpu3.cpp

```diff
@@ -6069,6 +6070,9 @@ TEST_F(NVFuserTest, FusionBroadcastPersistentReduction_CUDA) {
 // Repro for
 // https://github.com/csarofeen/pytorch/issues/2094
 TEST_F(NVFuserTest, FusionRepro2094_CUDA) {
+  // disable cast optimization, which causes numerical issue on tests
+  optimization::OptimizationPassGuard guard(
+      optimization::OptimizationPassCategory::PreSegmenter, false);
```

Comment:
@zasdfgbnm Here you go.
I also need to expose this on python side. There's a python test failing numerics

Reply:
What are the casts you want to optimize in https://github.com/NVIDIA/Fuser/issues/342? I think there are safe casts and unsafe casts, for example, `fp16->fp32->fp16` is safe, but `fp32->fp16->fp32` is unsafe. Similarly `int32 -> fp32 -> int32` is unsafe because of fp32 only has 24 bit of mantissa, but `int32 -> fp64 -> int32` is safe. I think for the simplification of safe casts, it is should always be on, but for unsafe one, it might be on or off, the default value for on/off depend on what user experience we want to deliver.

Is it possible to analyze based on safeness?

I think the principle should be like this: removing the intermediate is a noop, if and only if the intermediate contains strictly more or equal information than either the starting dtype or the target dtype. That is, removing the intermediate is a noop, if and only if the intermediate is not a bottleneck.

More formally, I think the set of dtypes is a [partially ordered set](https://en.wikipedia.org/wiki/Partially_ordered_set), and we can define
```
bool operator>=(PrimDataType dtype1, PrimDataType dtype2)
```
as having more or equal information, and it is safe if and only if `intermediate_dtype >= start_dtype || intermediate_dtype >= target_dtype`.

Reply:
I totally agree with you and I was gonna go down that path. Unfortunately the case I'm trying to patch in #335 is an unsafe case.
```
    T11 = fd.ops.cast(T9, dtype=DataType.Float)
    T12 = fd.ops.cast(T10, dtype=DataType.Float)
    T13 = fd.ops.add(T11, T12)
    T14 = fd.ops.cast(T13, dtype=DataType.Half)
    T15 = fd.ops.cast(T14, dtype=DataType.Float)
```
So it's casting `fp32->fp16->fp32`. So this is not a case where we are simply removing noop. Hence the reason to have the knob on switching it on/off.

Reply:
Oh, I didn't know the original case was not safe. Maybe we would need something like clang-tidy 😄

Reply:
I agree with having this pass, and I also agree enabling unsafe casts by default while providing a mechanism to disable it.

But on the side, why do we have such cast in the fusion definition in the first place? Is this something from user that we have no control of, or it is an artifact of our op definition? Is it possible to not have this cast?

For example, if we define our op `add` as:
```
TensorView* add(TensorView* tv1, TensorView* tv2) {
  auto math_type = DataType::Float;
  auto result_type = DataType::Half;
  tv1 = castOp(math_type, tv1);
  ....
  return castOp(result_type, result_in_math_type);
}
```
Then we should refactor our codebase to better handle type promotion. This doesn't sound easy, and I don't t think we must go along this path in the short term. But in the long term, besides having this pass, I think improving type promotion is also the direction to go.

Reply:
I think you already answered your question there. Yes this is likely coming from type promotion from our user. They are already working on their code to clean up the logic.
But like you said, we have a python API and user can go wild and define fusion as they say fit. Imagine another integration that uses nvfuser in their own parsing, that's beyond our control of how the fusion is defined. So having this path do seem like an unfortunate necessity.

So I think we are already on the same page in terms of what needs to be done and what is currently being done. Granted that the optimization pass itself here isn't bullet-proof. :stuck_out_tongue: But we can keep improving it in the future as we see fit.

Reply:
But you are also right about the type promotion logic in nvfuser C++ API. We could insert casts there as well that we might take another look at.

There are some of the type promotion that was there for historical reasons, which we might be able to improve. But generally speaking, if it's a combination of cast logic embedded by user + C++ API, it's tricky to get them to work seamlessly. Hence an optimization pass like this hopefully will clean up the mess.